### PR TITLE
Add return_edges argument to DataSet.assign_resolution_bins()

### LINKED ISSUE #184: rs.DataSet.assign_resolution_bins ought to return bin edges
The current bin labels assume formatting that takes up too much space. It would be better if this method just returned bin edges which users can treat however they want.

### PR DESCRIPTION
**Pull request by JBGreisman**
Fixes #184 by adding a `return_edges` argument to `DataSet.assign_resolution_bins()`. This PR adds that flexibility to the API, but otherwise does not change the default behavior of the function to preserve backwards compatibility.

### CODE REVIEW

File: reciprocalspaceship/utils/binning.py
Comment (kmdalton) [inline reciprocalspaceship/utils/binning.py:60]: If anyone has used this function in their code, this will change the default behavior. I'd move that if we're going to change the default behavior we may as well just get rid of the labels entirely at the `utils` level and migrate the format string stuff to the top level API. This has the side effect of letting the formatting be user-facing via the intended API.

File: reciprocalspaceship/dataset.py
Comment (kmdalton) [inline reciprocalspaceship/dataset.py:897]: I always find it hard to infer what "bin edges" means. For instance, is it an bins X 2 array or array with  bins + 1 entries? This needs to be clearer.

**Review by kmdalton — CHANGES_REQUESTED** [review]
I have suggested a slight refactor. I haven't gone over the tests very carefully yet. I want to wait for your feedback on my suggestion.

File: reciprocalspaceship/dataset.py
Comment (JBGreisman) [reply reciprocalspaceship/dataset.py:897]: I'll specify that it's a 1d array with bins+1 entries

File: reciprocalspaceship/utils/binning.py
Comment (JBGreisman) [reply reciprocalspaceship/utils/binning.py:60]: yep. I think it's fine to change the API in `utils` because it's not intended to be user-facing. I agree with you that the formatting should be handled in `DataSet.assign_resolution_bins()` instead of the `utils` function

**Comment by JBGreisman** [comment]
I made the requested refactors to `bin_by_percentile()` and the corresponding changes to `DataSet.assign_resolution_bins()` to preserve the user-facing API

**Review by kmdalton — CHANGES_REQUESTED** [review]

File: reciprocalspaceship/dataset.py
Comment (JBGreisman) [reply reciprocalspaceship/dataset.py:910]: yep -- I agree with both points and will update the order and add the fourth output

File: reciprocalspaceship/dataset.py
Comment (JBGreisman) [reply reciprocalspaceship/dataset.py:922]: sure -- no strong opinion from me.

File: reciprocalspaceship/stats/completeness.py
Comment (JBGreisman) [reply reciprocalspaceship/stats/completeness.py]: done

**Comment by JBGreisman** [comment]
@kmdalton I think this is good to go now, but let me know if you have any other thoughts

File: reciprocalspaceship/dataset.py
Comment (kmdalton) [inline reciprocalspaceship/dataset.py:883]: I think you should be able to supply an array of bin edges.

File: reciprocalspaceship/stats/completeness.py
Comment (kmdalton) [inline reciprocalspaceship/stats/completeness.py:6]: I think you can accomplish the same thing you are using `assign_with_binedges` to do by using `groupby(['H', 'K', 'L']).first()`. I'm not sure if that would be better or not. Just thinking out loud.

**Review by kmdalton — CHANGES_REQUESTED** [review]
I think `assign_resolution_bins` should take an array of bin edges. i think refls that fall outside the bin range should probably get dropped by default, but i am open to suggestions.

File: reciprocalspaceship/stats/completeness.py
Comment (JBGreisman) [reply reciprocalspaceship/stats/completeness.py:6]: I don't quite follow -- the `groupby()` call would return the first instances of each Miller index, but `assign_with_binedges` returns which resolution bin each reflection gets assigned to. In this case, the goal is to use the resolution bins determined for the provided `dataset` to assign the reflections in the 100% complete instance. 

If we make your above suggestion (to have `assign_resolution_bins()`) take bin edges, then we could refactor this to use that function instead of `utils`.

Comment (kmdalton) [reply reciprocalspaceship/stats/completeness.py:6]: okay we should probably do it via an `assign_resolution_bins` which accepts bin edges.

**Comment by JBGreisman** [comment]
> i think refls that fall outside the bin range should probably get dropped by default, but i am open to suggestions.

This is different than the current behavior in `rs.utils.assign_with_binedges()`, but could be implemented in `DataSet.assign_resolution_bins()`by pre-dropping any reflections before the call. If I do not do the pre-processing, the behavior will be to raise a `ValueError` that says `"This function assumes 'bin_edges' contain every entry in 'data'" `.

I hesitate to auto-drop reflections from the DataSet, but I think that could be acceptable as long we add a corresponding remark in the docstring

File: reciprocalspaceship/dataset.py
Comment (JBGreisman) [reply reciprocalspaceship/dataset.py:884]: New commits support this

**Comment by JBGreisman** [comment]
I updated this PR to allow `DataSet.assign_resolution_bins()` to support making assignments based on explicit bin edges. I made the following decisions regarding API:

- Reflections outside of provided bin edges are dropped
- Bin edges can be provided in ascending or descending order, and assignments will depend on that order

File: reciprocalspaceship/dataset.py
Comment (kmdalton) [inline reciprocalspaceship/dataset.py:930]: ideally you would not make a copy in the case of `inplace=True`.

File: tests/test_dataset_binning.py
Comment (kmdalton) [inline tests/test_dataset_binning.py:91]: can this be made to test for correctness? right now it only tests to see the result is not incorrect in one particular way.

File: reciprocalspaceship/dataset.py
Comment (kmdalton) [inline reciprocalspaceship/dataset.py:895]: good.

File: reciprocalspaceship/dataset.py
Comment (JBGreisman) [reply reciprocalspaceship/dataset.py:930]: While I agree in theory, I don't think it's possible to do so here without hitting a `SettingWithCopyWarning`. Let me know if you can think of a way to do so here -- I've tried all permutations I could think of.

File: tests/test_dataset_binning.py
Comment (JBGreisman) [reply tests/test_dataset_binning.py:91]: Sure -- this tests correctness against the assignments made using `assign_resolution_bins(2)`, but I can also check against `dHKL`, explicitly

Comment (JBGreisman) [reply tests/test_dataset_binning.py:91]: yep -- just updated this to test explicitly against `(1 - bin)` in the reverse case

File: reciprocalspaceship/dataset.py
Comment (JBGreisman) [reply reciprocalspaceship/dataset.py:930]: nvm -- I _think_ I got it

Comment (JBGreisman) [reply reciprocalspaceship/dataset.py]: resolved with new commits

**Review by kmdalton — APPROVED** [review]
looks good to me. great stuff!

**Pull request merged**
